File: src/main/java/io/bloco/faker/helpers/StringHelper.java
package io.bloco.faker.helpers;

import java.text.Normalizer;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class StringHelper {

    public interface StringReplacer {
        String replaceWith(Matcher matcher);
    }

    public String replaceMethod(String input, String regex, StringReplacer replacer) {
        Pattern pattern = Pattern.compile(regex);
        Matcher matcher = pattern.matcher(input);
        StringBuffer stringBuffer = new StringBuffer();
        while (matcher.find()) {
            String replacement = replacer.replaceWith(matcher);
            matcher.appendReplacement(stringBuffer, replacement);
        }
        matcher.appendTail(stringBuffer);
        return stringBuffer.toString();
    }

    public String snakeToCamel(String input) {
        return replaceMethod(input, "_(\\p{Lower})", new StringReplacer() {
            @Override
            public String replaceWith(Matcher matcher) {
                String letter = matcher.group(1);
                return letter.toUpperCase();
            }
        });
    }

    public String camelToSnake(String input) {
        return replaceMethod(input, "(?<=\\w)(\\p{Upper})", new StringReplacer() {
            @Override
            public String replaceWith(Matcher matcher) {
                return "_" + matcher.group(1);
            }
        }).toLowerCase();
    }

    public String join(List<String> list, String separator) {
        StringBuilder sb = new StringBuilder();
        boolean first = true;
        for (String item : list) {
            if (first)
                first = false;
            else
                sb.append(separator);
            sb.append(item);
        }
        return sb.toString();
    }

    public String normalize(String input) {
        String normalized = Normalizer.normalize(input, Normalizer.Form.NFD); // Separate glyphs
        normalized = normalized.replaceAll("[^\\p{ASCII}]", ""); // Remove glyphs
        normalized = normalized.replaceAll("\\W", ""); // Remove anything but letters and numbers
        return normalized.toLowerCase();
    }
}

File: lib/src/main/kotlin/io/bloco/faker/helpers/StringHelper.kt
package io.bloco.faker.helpers

import java.text.Normalizer
import java.util.*
import java.util.regex.Pattern

fun String.snakeToCamel(): String {
    return this.replace("_(\\p{Lower})".toRegex()) { matcher ->
        matcher.groupValues[1].uppercase(Locale.getDefault())
    }
}

fun String.camelToSnake(): String {
    return this.replace("(?<=\\w)(\\p{Upper})".toRegex()) { matcher ->
        "_" + matcher.groupValues[1].lowercase(Locale.getDefault())
    }.lowercase(Locale.getDefault())
}

fun String.normalize(): String {
    val asciiPattern = Pattern.compile("[^\\p{ASCII}]")
    val nonWordPattern = Pattern.compile("\\W")

    var normalized = Normalizer.normalize(this, Normalizer.Form.NFD) // Separate glyphs
    normalized = asciiPattern.matcher(normalized).replaceAll("") // Remove glyphs
    normalized = nonWordPattern.matcher(normalized)
        .replaceAll("") // Remove anything but letters and numbers
    return normalized.lowercase(Locale.getDefault())
}
